Add setting items to control items in status bar

TerminalFi opened the pull request:
Release Notes:

Added settings

- `show_status_bar` for hiding the status bar
- `show_active_language` for hiding the active language in the status bar
- `show_cursor_position` for hiding the cursor position
- `show_feedback_icon` for hiding the share feedback button

Added `button` properties to other status bar items to hide.

<del>
- Added new settings object that controls various elements in status bar


```json
  "status_bar": {
    "visible": true,
    "elements": {
      "project": false,
      "assistant": false,
      "feedback": false,
      "terminal": false,
      "copilot": false,
      "cursor_position": false,
      "language_selector": false,
      "chat": false,
      "notification": false,
      "collaboration": false,
      "diagnostics": false
    }
  }
```

The only thing I do not like is that `assistant_panel` `copilot`, etc all have settings like the below

```
"assistant": {
    "button": false
  },
```

This seems like now we would have 2 different places to contain status bar items
</del>

## Preview


https://github.com/zed-industries/zed/assets/32599364/e605d6bd-6c9b-442b-bc8b-4591ce5796f6



closes https://github.com/zed-industries/zed/issues/8169

TerminalFi committed:
adding settings to control status bar items

TerminalFi commented:
@cla-bot check

TerminalFi commented:
The other options is to call this UiSettings

And then we can merge any UI element through this setting object. Such as title bar, etc

slymax commented:
What about a structure like this:

```json
"status_bar": {
  "visible": true,
  "elements": {
    "assistant": true,
    "feedback": true
  }
}
```

and yeah I think the duplicate settings should be removed before this gets added

wmstack commented:
Hmm any way to order the elements?

TerminalFi committed:
prep workspace settings

TerminalFi committed:
update collab ui

TerminalFi committed:
update assistant panel references

support disabling status bar

remaining status bar items

default settings

remove Option

area specific button settings

TerminalFi committed:
update assistant panel references

TerminalFi committed:
support disabling status bar

TerminalFi committed:
remaining status bar items

TerminalFi committed:
default settings

slymax commented:
> Hmm any way to order the elements?

good question. I think someone from the team should chime on whether the order should be customizable

slymax commented on crates/diagnostics/src/items.rs:
```diff
@@ -22,6 +23,17 @@ pub struct DiagnosticIndicator {
 
 impl Render for DiagnosticIndicator {
     fn render(&mut self, cx: &mut ViewContext<Self>) -> impl IntoElement {
+        if let Some(elements) = WorkspaceSettings::get_global(cx)
+            .status_bar
+            .clone()
+            .elements
+        {
+            // Now that we have elements, check cursor_position
+            if !elements.diagnostics.unwrap_or(true) {
+                // If cursor_position is true or None (defaulting to true)
+                return div();
+            }
+        }
```
not a rust expert myself, unfortunately

ejjonny replied:
Should these be optional? Is that how other settings are implemented? If so, it feels like the fallback setting should come from a proper source in the case that they are `None` rather than dispersed `unwrap_or` calls.

TerminalFi replied:
You are right, others generally aren't optional. They are just bool instead. 

I can make this change

probablykasper commented:
Seems fair to have the order customizable.
```json
"status_bar": {
  "visible": true,
  "elements": [
    "assistant",
    "feedback",
    "etc"
  ]
}
```

TerminalFi commented:
Just stating, this is outside the scope of what I am working on. As reading the code, I elieve it would require some rework.

> Seems fair to have the order customizable.
> 
> ```json
> "status_bar": {
>   "visible": true,
>   "elements": [
>     "assistant",
>     "feedback",
>     "etc"
>   ]
> }
> ```

this doesn't handle if it is left or right side of the status bar. Which is also needed. I think this would be out of scope for this PR. Is there a issue for this already?

maxdeviant commented on crates/copilot_ui/src/copilot_button.rs:
```diff
@@ -38,6 +38,18 @@ pub struct CopilotButton {
+        {
+            // Now that we have elements, check cursor_position
+            if !elements.copilot.unwrap_or(true) {
+                // If cursor_position is true or None (defaulting to true)
```
Do we need this comment? It seems to be restating what the code is doing (and it also seems incorrect).

maxdeviant commented on crates/quick_action_bar/src/quick_action_bar.rs:
```diff
@@ -121,15 +121,22 @@ impl Render for QuickActionBar {
             },
         );
+        let mut assistant_icon = true;
+        if let Some(elements) = WorkspaceSettings::get_global(cx)
```
It would be better to leave this expressed in the current structure using `.when`.

maxdeviant reviewed (changes requested):
I'm not sure combining these under a single `status_bar` config is the way to go.

I think we might want to keep them nested under their respective sections, so that each component is somewhat independent.

This will be more important in the future when extensions expose their own settings.

TerminalFi commented:
> I'm not sure combining these under a single `status_bar` config is the way to go.
> 
> 
> 
> I think we might want to keep them nested under their respective sections, so that each component is somewhat independent.
> 
> 
> 
> This will be more important in the future when extensions expose their own settings.


Sure, I can take another stab at it. Moving things back under specific scopes

What able for the general status-bar ?

Should it still be status_bar: { visible: true } or just a single top level key value like show_status_bar: true

TerminalFi committed:
Merge remote-tracking branch 'origin/status_bar_settings' into status_bar_settings

TerminalFi committed:
fix missing keys

TerminalFi committed:
Merge branch 'main' into status_bar_settings

TerminalFi committed:
update to when statements

TerminalFi committed:
additional logic refactoring

TerminalFi committed:
attempt copilot hiding

TerminalFi commented:
@maxdeviant The only thing I can't figure out is hiding the Copilot icon. The code I put doesn't seem to work... any ideas?

TerminalFi commented:
Could probably put cursor position and active language in editor settings

maxdeviant commented:
Apologies for the radio silence here.

I think the surface area for this PR is a bit too large at the moment. Maybe we could try opening up smaller PRs for each individual new setting?

TerminalFi commented:
> Apologies for the radio silence here.
> 
> 
> 
> I think the surface area for this PR is a bit too large at the moment. Maybe we could try opening up smaller PRs for each individual new setting?

Sure!
